main.cpp

```diff
@@ -180,139 +123,268 @@ int main()
 
 #include <iostream>
 
+struct DoubleType;
+struct IntType;
+
 struct FloatType
 {
-    float add(float lhs, float rhs);
-    float subtract(float lhs,float rhs );
-    float multiply(float lhs, float rhs );
-    float divide(float lhs,float rhs );
+    float value;
+
+    FloatType& add( float rhs );
+    FloatType& subtract( float rhs );
+    FloatType& multiply( float rhs );
+    FloatType& divide( float rhs );
+
+    FloatType& add(const DoubleType& dt);
+    FloatType& subtract(const DoubleType& dt);
+    FloatType& multiply(const DoubleType& dt);
+    FloatType& divide(const DoubleType& dt);
+
+    FloatType& add(const IntType& dt);
+    FloatType& subtract(const IntType& dt);
+    FloatType& multiply(const IntType& dt);
+    FloatType& divide(const IntType& dt);
+
+    FloatType& add(const FloatType& dt);
+    FloatType& subtract(const FloatType& dt);
+    FloatType& multiply(const FloatType& dt);
+    FloatType& divide(const FloatType& dt);
+
+    FloatType(float v) : value(v) {}
 };
 
-float FloatType::add(float lhs, float rhs)
+FloatType& FloatType::add(float rhs)
 {
-    return lhs + rhs;
+    value += rhs;
+    return *this;
 }
 
-float FloatType::subtract(float lhs, float rhs)
+FloatType& FloatType::subtract(float rhs)
 {
-    return lhs - rhs;
+    value -= rhs;
+    return *this;
 }
 
-float FloatType::multiply(float lhs, float rhs)
+FloatType& FloatType::multiply(float rhs)
 {
-    return lhs * rhs;
+    value *= rhs;
+    return *this;
 }
 
-float FloatType::divide(float lhs, float rhs)
+FloatType& FloatType::divide(float rhs)
 {
     // warn the user if they're doing floating-point-division-by-zero 
     // but not prevent the division from happening
     if (rhs == 0.f) 
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";        
-    return lhs / rhs;
+        std::cout << "warning: floating point division by zero!\n";        
+    value /= rhs;
+    return *this;
 }
 
 struct DoubleType
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs,double rhs );
-    double multiply(double lhs, double rhs );
-    double divide(double lhs,double rhs );
+    double value;
+
```

Comment:
This does not demonstrate the following:
```
Edit your 3 structs so that they own a heap-allocated primitive type 
without using smart pointers named 'value'
```
- [ ] revise